### PR TITLE
Bagfix and migration phase 2

### PR DESCRIPTION
**Pull request by nshchego**
### Details:
 - *item1*
 - *...*

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp [hunk 1]
```diff
@@ -84,6 +89,14 @@ void MKLDNNInputNode::initSupportedPrimitiveDescriptors() {
         auto mem_tdesc = MKLDNNMemoryDesc(getChildEdgeAt(0)->getDims(), outputDataType);
         dataConfig.desc = mem_tdesc;
         config.outConfs.push_back(dataConfig);
+        // ReadValue operation expects constant input
+        if (!getParentEdges().empty()) {
```
[the remark this answers is not in the document]
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_input_node.cpp:93]: I prefer to leave general case,

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_def_conv_node.cpp [hunk 2]
```diff
@@ -741,65 +739,80 @@ struct jit_uni_def_conv_kernel_f32 : public jit_uni_def_conv_kernel, public jit_
...
+    if (getParentEdges().size() != 3)
+        IE_THROW() << errorPrefix << "has incorrect number of input edges";
+    if (getChildEdges().empty())
+        IE_THROW() << errorPrefix << "has incorrect number of output edges";
+
+    if (getParentEdgeAt(0)->getDims().ndims() != 4) {
+        IE_THROW() << "Deformable convolution layer. Unsupported mode. Only 4D blobs are supported as input.";
+    }
+
+    if (getParentEdgeAt(0)->getDims().ndims() != 4) {
```
[the remark this answers is not in the document]
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_def_conv_node.cpp]: FIxed

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_def_conv_node.cpp [hunk 3]
```diff
@@ -741,65 +739,80 @@ struct jit_uni_def_conv_kernel_f32 : public jit_uni_def_conv_kernel, public jit_
...
+
+    if (getParentEdgeAt(2)->getDims().ndims() != 4) {
+        IE_THROW() << errorPrefix << "doesn't support 2nd input with rank: " << getParentEdgeAt(2)->getDims().ndims();
+    }
+
+    if (getChildEdgeAt(0)->getDims().ndims() != 4) {
+        IE_THROW() << errorPrefix << "doesn't support output with rank: " << getChildEdgeAt(0)->getDims().ndims();
+    }
+
+    bool isMerged = (!getMergeWith().empty());
```
[the remark this answers is not in the document]
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_def_conv_node.cpp]: Fixed

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_memory_node.cpp [hunk 4]
```diff
@@ -31,7 +58,7 @@ void MKLDNNMemoryOutputNode::initSupportedPrimitiveDescriptors() {
     if (!supportedPrimitiveDescriptors.empty())
         return;
 
-    InferenceEngine::Precision precision = getCnnLayer()->insData[0].lock()->getPrecision();
+    InferenceEngine::Precision precision = getOriginalInputPrecisions()[0];
```
[the remark this answers is not in the document]
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_memory_node.cpp]: Fixed

**Comment by mandrono** [comment]
https://github.com/dmitry-gorokhov/openvino/pull/45/files#diff-0ec2c2a8d308ceb06e9f0d16319099aa1bb461c49a19e51c5267ced413f25f77R46-R64

**Comment by mandrono** [comment]
@nshchego also enable this please

https://github.com/dmitry-gorokhov/openvino/blob/1505bf04277b6ec58e8ecb7755638a5a8fa8a5ca/inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp#L46-L64

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_split_node.cpp [hunk 5]
```diff
@@ -80,6 +87,15 @@ MKLDNNSplitNode::MKLDNNSplitNode(const std::shared_ptr<ngraph::Node>& op, const
     if (!isSupportedOperation(op, errorMessage)) {
         IE_THROW(NotImplemented) << errorMessage;
     }
+
+    if (ngraph::as_type_ptr<const ngraph::op::v1::Split>(op)) {
+        algorithm = SplitDefault;
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_split_node.cpp:92]: If this algorithm is not used anyhow I would suggest to remove it from enum.

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/convolution.cpp [hunk 6]
```diff
@@ -102,6 +102,33 @@ INSTANTIATE_TEST_CASE_P(
         ::testing::Values(CommonTestUtils::DEVICE_CPU)),
     ConvolutionLayerTest::getTestCaseName);
 
+namespace specificWeightLayout {
```
Comment (dmitry-gorokhov) [inline inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/convolution.cpp:105]: Could you please add more detailed description why we need this test into the comment?

File: inference-engine/tests/functional/plugin/cpu/test_utils/node_types.hpp [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    groupConvolution,
+    groupConvolutionBackpropData
+};
+
+inline std::string nodeType2PluginType(nodeType nt) {
+    if (nt == nodeType::convolution) return "Convolution";
+    if (nt == nodeType::convolutionBackpropData) return "Deconvolution";
+    if (nt == nodeType::groupConvolution) return "Convolution";
+    if (nt == nodeType::groupConvolutionBackpropData) return "Deconvolution";
+    assert(!"unknown node type");
```
Comment (dmitry-gorokhov) [inline inference-engine/tests/functional/plugin/cpu/test_utils/node_types.hpp:22]: Maybe just throw exception?

File: inference-engine/tests/functional/plugin/cpu/test_utils/node_types.hpp [hunk 8]
```diff
@@ -0,0 +1,33 @@
+// Copyright (C) 2021 Intel Corporation
```
Comment (dmitry-gorokhov) [inline inference-engine/tests/functional/plugin/cpu/test_utils/node_types.hpp:1]: Maybe just move into cpu_test_utils.hpp?

File: inference-engine/include/ie_common.h [hunk 9]
```diff
@@ -483,7 +483,7 @@ struct NullStream {
     auto res = actual->function(&resp);                                                                     \
     if (res != OK) IE_EXCEPTION_SWITCH(res, ExceptionType,                                                  \
             InferenceEngine::details::ThrowNow<ExceptionType>{}                                             \
-                <<= std::stringstream{} << IE_LOCATION)
+                <<= std::stringstream{} << IE_LOCATION << resp.msg)
```
[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/include/ie_common.h:486]: without this change in error message reported only about that error was occur in infer request
but we have more detail description
Like: [NOT_IMPLEMENTED] Unsupported operation of type: Gather name: Gather_4121
Cannot fallback on ngraph reference implementation (Ngraph::Node::evaluate() is not implemented)

File: inference-engine/tests/functional/plugin/cpu/test_utils/node_types.hpp (hunk 8 above)
Comment (mandrono) [reply inference-engine/tests/functional/plugin/cpu/test_utils/node_types.hpp]: moved

File: inference-engine/tests/functional/plugin/cpu/test_utils/node_types.hpp (hunk 7 above)
Comment (mandrono) [reply inference-engine/tests/functional/plugin/cpu/test_utils/node_types.hpp]: fixed

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/convolution.cpp (hunk 6 above)
Comment (mandrono) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/convolution.cpp:107]: description added

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_split_node.cpp (hunk 5 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_split_node.cpp]: removed

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_memory_node.cpp [hunk 10]
```diff
@@ -50,8 +77,26 @@ void MKLDNNMemoryOutputNode::execute(mkldnn::stream strm)  {
     inputMemoryNode->storeState(srcMemory);
 }
 
+bool MKLDNNMemoryInputNode::isSupportedOperation(const std::shared_ptr<const ngraph::Node>& op, std::string& errorMessage) noexcept {
+    try {
+        if (!MKLDNNPlugin::one_of(op->get_type_info(),
+                ngraph::op::v3::ReadValue::type_info,
+                ngraph::op::v6::ReadValue::type_info)) {
+            errorMessage = "Node is not an instance of ReadValue.";
```
[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_memory_node.cpp]: added

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_memory_node.cpp [hunk 11]
```diff
@@ -7,15 +7,42 @@
...
+        _id = readValueOp->get_variable_id();
+    }
+}
+
+bool MKLDNNMemoryOutputNode::isSupportedOperation(const std::shared_ptr<const ngraph::Node>& op, std::string& errorMessage) noexcept {
+    try {
+        if (!MKLDNNPlugin::one_of(op->get_type_info(),
+                ngraph::op::v3::Assign::type_info,
+                ngraph::op::v6::Assign::type_info)) {
+            errorMessage = "Node is not an instance of Assign.";
```
[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_memory_node.cpp]: added

File: inference-engine/src/mkldnn_plugin/ngraph_transformations/convert_tile_to_seq_tiles.cpp [hunk 12]
```diff
@@ -43,8 +43,10 @@ MKLDNNPlugin::ConvertTileToSeqTiles::ConvertTileToSeqTiles() {
             }
         }
 
-        if (num_of_tile_dims < 2)
-            return false;
+        if (num_of_tile_dims == 0) {
```
[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/ngraph_transformations/convert_tile_to_seq_tiles.cpp:46]: Tile node was updated

File: inference-engine/src/mkldnn_plugin/mkldnn_node.cpp [hunk 13]
```diff
@@ -1343,7 +1335,15 @@ bool MKLDNNNode::canBePerformedAsScaleShift(const MKLDNNNode *parentNode) const
         for (size_t i = 0; i < getParentEdges().size(); i++) {
             if (i == fusingPort)
                 continue;
-            if (!isPerTensorOrPerChannelBroadcastable(dataShape, getParentEdgeAt(i)->getDims().ToSizeVector()))
+            auto weightShape = getParentEdgeAt(i)->getDims().ToSizeVector();
+            // WA: waiting fix from ngraph
```
[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/mkldnn_node.cpp]: Issue: 51993
It was incorrect comment, I am waiting spec update to fix PRelu fully

File: inference-engine/src/mkldnn_plugin/mkldnn_memory.cpp [hunk 14]
```diff
@@ -701,6 +693,19 @@ MKLDNNMemoryDesc::operator InferenceEngine::TensorDesc() const {
     for (int i = 0; i < inner_ndims; i++) {
         total_block_per_dim[blk_desc.inner_idxs[i]] *= blk_desc.inner_blks[i];
     }
+    std::vector<size_t> outer_block_dims(std::begin(dims), std::begin(dims) + outer_ndims);
```
[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/mkldnn_memory.cpp:696]: tests was added

File: inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp [hunk 15]
```diff
@@ -55,6 +55,7 @@ void MKLDNNGraphOptimizer::ApplyCommonGraphOptimizations(MKLDNNGraph &graph) {
 // TODO [NM]: transformation should be implemented w/o using of CNNLayer
 //    MergeTwoEqualScaleShifts(graph);
 //    graph.RemoveDroppedNodes();
+    Reshape1DConvolution(graph);
```
[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp]: reimplemented via TypeRelaxed

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp [hunk 16]
```diff
@@ -217,13 +218,10 @@ void MKLDNNGraph::Replicate(const CNNNetwork &network, const MKLDNNExtensionMana
...
     std::deque<ngraph::Output<ngraph::Node>> unusedOutputs;  // nodes which has no consumers (output or just unused)
 
-    auto getParentPort = [](const std::shared_ptr<ngraph::Node> op, const std::shared_ptr<ngraph::Node> parentOp) -> int {
-        // TODO [NM]: do we have a better way how to determine parent port?
-        for (size_t port = 0; port < parentOp->get_output_size(); port++) {
-            auto outputs = parentOp->get_output_target_inputs(port);
-            for (auto& output : outputs) {
-                if (op == output.get_node()->shared_from_this())
-                    return static_cast<int>(port);
+    auto getParentPort = [](const std::shared_ptr<ngraph::Node> op, const std::shared_ptr<ngraph::Node> parentOp, const size_t port) -> int {
```
[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp]: changed

File: inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp [hunk 17]
```diff
@@ -251,15 +240,16 @@ void MKLDNNGraph::Replicate(const CNNNetwork &network, const MKLDNNExtensionMana
...
 
         if (op->get_type_info() == ngraph::op::v0::Result::type_info) {
-            auto prev = op->get_input_node_shared_ptr(0);
-            std::string inputID;
-            inputID = prev->get_friendly_name();
-            if (prev->get_output_size() > 1) {
-                inputID += "." + std::to_string(op->get_input_source_output(0).get_index());
+            const auto &input = op->input_value(0);
+            NGRAPH_SUPPRESS_DEPRECATED_START
+            auto name = input.get_tensor().get_name();
```
[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp:247]: TODO was added

**Comment by nshchego** [comment]
> @nshchego also enable this please
> 
> https://github.com/dmitry-gorokhov/openvino/blob/1505bf04277b6ec58e8ecb7755638a5a8fa8a5ca/inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp#L46-L64

Done

**Comment by mandrono** [comment]
> > @nshchego also enable this please
> > https://github.com/dmitry-gorokhov/openvino/blob/1505bf04277b6ec58e8ecb7755638a5a8fa8a5ca/inference-engine/src/mkldnn_plugin/mkldnn_infer_request.cpp#L46-L64
> 
> Done

remove ```THROW_IE_EXCEPTION << "[NM] Not implemented";``` please

File: inference-engine/src/mkldnn_plugin/ngraph_transformations/reshape_1d_ops.cpp [hunk 18]
```diff
@@ -28,15 +29,38 @@ std::shared_ptr<ngraph::Node> convert(const ngraph::Output<ngraph::Node> & data,
...
 
-    return std::make_shared<T>(data,
-                               weights,
-                               new_strides,
-                               new_pads_begin,
-                               new_pad_end,
-                               new_dilations,
-                               node->get_auto_pad());
+    if (std::dynamic_pointer_cast<ngraph::op::TypeRelaxedBase>(node)) {
+        std::vector<ngraph::element::Type> inputPrecisions;
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/ngraph_transformations/reshape_1d_ops.cpp:36]: seems like unsued?

File: inference-engine/src/mkldnn_plugin/ngraph_transformations/reshape_1d_ops.cpp [hunk 19]
```diff
@@ -28,15 +29,38 @@ std::shared_ptr<ngraph::Node> convert(const ngraph::Output<ngraph::Node> & data,
...
+        for (auto& inputs : node->inputs()) {
+            inputPrecisions.push_back(inputs.get_element_type());
+        }
+
+        std::vector<ngraph::element::Type> outputPrecisions;
+        for (auto& output : node->outputs()) {
+            outputPrecisions.push_back(output.get_element_type());
+        }
+
+        return std::make_shared<ngraph::op::TypeRelaxed<BaseOp>>(std::vector<ngraph::element::Type>{ngraph::element::f32, ngraph::element::f32},
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/ngraph_transformations/reshape_1d_ops.cpp:46]: just didn't get why always fp32 here?

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_conv_node.cpp [hunk 20]
```diff
@@ -129,6 +108,30 @@ void MKLDNNConvolutionNode::getSupportedDescriptors() {
     if (!descs.empty())
         return;
 
+    if (getAlgorithm() == ConvolutionCommon) {
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_conv_node.cpp:111]: Seems like this change is not actual anymore?

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_conv_node.h [hunk 21]
```diff
@@ -44,6 +44,16 @@ class MKLDNNConvolutionNode : public MKLDNNNode {
     std::vector<float> weightsZeroPoints;
     std::vector<int32_t> outputCompensation;
 
+    void setStride(const std::vector<ptrdiff_t> &_stride) { stride = _stride; }
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_conv_node.h:47]: Seems like this change is not actual anymore?

File: inference-engine/src/mkldnn_plugin/ngraph_transformations/reshape_1d_ops.cpp (hunk 18 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/ngraph_transformations/reshape_1d_ops.cpp]: removed

File: inference-engine/src/mkldnn_plugin/ngraph_transformations/reshape_1d_ops.cpp (hunk 19 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/ngraph_transformations/reshape_1d_ops.cpp:36]: it's temporary precisions to create Convolution, after creation these precisions will be changed on original

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_conv_node.cpp (hunk 20 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_conv_node.cpp]: removed

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_conv_node.h (hunk 21 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_conv_node.h]: removed

**Pull request merged**